### ISSUE TITLE
[bitnami/kubectl] - CVE-2021-46848 and CVE-2022-32221

### ISSUE THREAD

**Issue by fryz** [bitnami/containers#19192]
### Name and Version

bitnami/kubectl:1.22-debian-11

## CVE-2022-32221

```
$ trivy image bitnami/kubectl:1.22-debian-11 | grep CVE-2022-32221
2023-01-05T12:11:24.235-0500    INFO    Vulnerability scanning is enabled
2023-01-05T12:11:24.235-0500    INFO    Secret scanning is enabled
2023-01-05T12:11:24.235-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-05T12:11:24.235-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-05T12:11:31.297-0500    INFO    Detected OS: debian
2023-01-05T12:11:31.298-0500    INFO    Detecting Debian vulnerabilities...
2023-01-05T12:11:31.313-0500    INFO    Number of language-specific files: 0
│ libcurl3-gnutls   │ CVE-2022-32221   │ CRITICAL │ 7.74.0-1.3+deb11u3    │                  │ curl: POST following PUT confusion                           

$ trivy image debian:bullseye | grep CVE-2022-32221
2023-01-05T12:13:01.447-0500    INFO    Vulnerability scanning is enabled
2023-01-05T12:13:01.447-0500    INFO    Secret scanning is enabled
2023-01-05T12:13:01.447-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-05T12:13:01.447-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-05T12:13:06.972-0500    INFO    Detected OS: debian
2023-01-05T12:13:06.972-0500    INFO    Detecting Debian vulnerabilities...
2023-01-05T12:13:06.978-0500    INFO    Number of language-specific files: 0
```

## CVE-2021-46848

```

@carrodher,

I'd like to reopen this please, as the latest debian-11 kubectl 1.22 image is still showing vulnerable to this CVE, but I've confirmed (using the same trivy commands you ran above) that the CVE has been fixed in the upstream debian images:

Can we get this image rebuilt please?

Thanks!

[14:20:31]$ trivy image debian:bullseye | grep CVE-2021-46848
2023-01-04T14:20:32.982-0500    INFO    Vulnerability scanning is enabled
2023-01-04T14:20:32.982-0500    INFO    Secret scanning is enabled
2023-01-04T14:20:32.982-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-04T14:20:32.982-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-04T14:20:38.884-0500    INFO    Detected OS: debian
2023-01-04T14:20:38.884-0500    INFO    Detecting Debian vulnerabilities...
2023-01-04T14:20:38.899-0500    INFO    Number of language-specific files: 0

[14:20:38]$ trivy image debian:bullseye-slim | grep CVE-2021-46848
2023-01-04T14:20:44.915-0500    INFO    Vulnerability scanning is enabled
2023-01-04T14:20:44.915-0500    INFO    Secret scanning is enabled
2023-01-04T14:20:44.915-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-04T14:20:44.915-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-04T14:20:49.278-0500    INFO    Detected OS: debian
2023-01-04T14:20:49.278-0500    INFO    Detecting Debian vulnerabilities...
2023-01-04T14:20:49.283-0500    INFO    Number of language-specific files: 0

[14:23:10]$ trivy image bitnami/kubectl:1.22-debian-11 | grep CVE-2021-46848
2023-01-04T14:23:21.043-0500    INFO    Vulnerability scanning is enabled
2023-01-04T14:23:21.043-0500    INFO    Secret scanning is enabled
2023-01-04T14:23:21.043-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-04T14:23:21.043-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-04T14:23:21.686-0500    INFO    Detected OS: debian
2023-01-04T14:23:21.686-0500    INFO    Detecting Debian vulnerabilities...
2023-01-04T14:23:21.714-0500    INFO    Number of language-specific files: 0
│ libtasn1-6        │ CVE-2021-46848   │ CRITICAL │ 4.16.0-2              │ 4.16.0-2+deb11u1 │ libtasn1: Out-of-bound access in ETYPE_OK                    │
```

### What is the expected behavior?

Because the aforementioned vulnerabilities are fixed in the upstream debian repositories, the bitnami/kubectl images shouldn't show as vulnerable. 

### What do you see instead?

Images are vulnerable. 

### Additional information

Opening a new issue following up on https://github.com/bitnami/containers/issues/12457, as I don't see a way to re-open and it wasn't clear if the closed issue would get attention.

**Comment by carrodher** [comment]
Hi, unfortunately, those security vulnerabilities are not fixed by the OS or the application itself, so although we built the images on a regular basis to provide the latest version of system packages, this kind of CVE will be reported while there is no new version patching the issue in the OS or the application. [Here](https://docs.bitnami.com/kubernetes/open-cve-policy/) you can find more info about this topic.

```console
$ trivy image --ignore-unfixed bitnami/kubectl
2023-01-07T17:46:39.846+0100	INFO	Vulnerability scanning is enabled
2023-01-07T17:46:39.846+0100	INFO	Secret scanning is enabled
2023-01-07T17:46:39.846+0100	INFO	If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-07T17:46:39.846+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.35/docs/secret/scanning/#recommendation for faster secret detection
2023-01-07T17:46:41.673+0100	INFO	Detected OS: debian
2023-01-07T17:46:41.673+0100	INFO	Detecting Debian vulnerabilities...
2023-01-07T17:46:41.688+0100	INFO	Number of language-specific files: 1
2023-01-07T17:46:41.688+0100	INFO	Detecting gobinary vulnerabilities...

bitnami/kubectl (debian 11.6)

Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


opt/bitnami/kubectl/bin/kubectl (gobinary)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

┌──────────────────┬────────────────┬──────────┬──────────────────────────────────────┬───────────────┬──────────────────────────────────────────────────────────────┐
│     Library      │ Vulnerability  │ Severity │          Installed Version           │ Fixed Version │                            Title                             │
├──────────────────┼────────────────┼──────────┼──────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────┤
│ golang.org/x/net │ CVE-2022-41717 │ MEDIUM   │ v0.3.1-0.20221206200815-1e63c2f08a10 │ 0.4.0         │ An attacker can cause excessive memory growth in a Go server │
│                  │                │          │                                      │               │ accepting...                                                 │
│                  │                │          │                                      │               │ https://avd.aquasec.com/nvd/cve-2022-41717                   │
└──────────────────┴────────────────┴──────────┴──────────────────────────────────────┴───────────────┴──────────────────────────────────────────────────────────────┘
```

The Bitnami Application Catalog (OpenSource) is based on Debian 11 but Bitnami, as part of VMware, provides a custom container and Helm Charts catalog based on the desired base image (generic distros such as Debian 10 & 11, CentOS 7, PhotonOS 3 & 4, Ubuntu 18.04, 20.04 & 22.04, RedHat UBI 8 & 9, or custom golden image) through the [VMware Tanzu Application Catalog](https://tanzu.vmware.com/content/blog/use-bitnami-apps-with-vmware-application-catalog).

**Comment by fryz** [comment]
Hey @carrodher - 

Thanks for the context and letting me know about 1.22 no longer being supported. 

Though I don't understand what you mean when you say the OS/application hasn't fixed the vulnerability. From what I can tell: 
1. Scanning the `debian:bullseye` and `debian:bullseye-slim` images show that the vulnerabilities in the title are not present (see the scan output in the description) 
2. The Debian Bug Tracker shows that both vulnerabilities have been fixed on `Bullseye`:
    * https://security-tracker.debian.org/tracker/CVE-2022-23221
    * https://security-tracker.debian.org/tracker/CVE-2021-46848

You mentioned:
> The Bitnami Application Catalog (OpenSource) is based on Debian 11 but Bitnami, as part of VMware, provides a custom container and Helm Charts catalog based on the desired base image (generic distros such as Debian 10 & 11, CentOS 7, PhotonOS 3 & 4, Ubuntu 18.04, 20.04 & 22.04, RedHat UBI 8 & 9, or custom golden image) through the [VMware Tanzu Application Catalog](https://tanzu.vmware.com/content/blog/use-bitnami-apps-with-vmware-application-catalog).

Does this mean that the image in the Tanzu catalog hasn't been patched? If so, do you have an ETA on when that might happen? 

Thanks!

**Comment by carrodher** [comment]
> Though I don't understand what you mean when you say the OS/application hasn't fixed the vulnerability. From what I can tell:
> 
> 1. Scanning the `debian:bullseye` and `debian:bullseye-slim` images show that the vulnerabilities in the title are not present (see the scan output in the description)
> 2. The Debian Bug Tracker shows that both vulnerabilities have been fixed on `Bullseye`:
>    
>    * https://security-tracker.debian.org/tracker/CVE-2022-23221
>    * https://security-tracker.debian.org/tracker/CVE-2021-46848

The base image used for the Bitnami containers is [`bitnami/minideb:bullseye`](https://github.com/bitnami/minideb) no `debian:bullseye`, but in any case, the mentioned CVEs are not present in the base image:
```console
$ trivy image --ignore-unfixed bitnami/minideb | grep 'CVE-2021-46848'
2023-01-09T20:54:13.942+0100	INFO	Vulnerability scanning is enabled
2023-01-09T20:54:13.942+0100	INFO	Secret scanning is enabled
2023-01-09T20:54:13.942+0100	INFO	If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-09T20:54:13.942+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.36/docs/secret/scanning/#recommendation for faster secret detection
2023-01-09T20:54:16.071+0100	INFO	Detected OS: debian
2023-01-09T20:54:16.071+0100	INFO	Detecting Debian vulnerabilities...
2023-01-09T20:54:16.080+0100	INFO	Number of language-specific files: 0

$ trivy image --ignore-unfixed bitnami/minideb | grep 'CVE-2022-32221'
2023-01-09T20:54:22.669+0100	INFO	Vulnerability scanning is enabled
2023-01-09T20:54:22.669+0100	INFO	Secret scanning is enabled
2023-01-09T20:54:22.669+0100	INFO	If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-09T20:54:22.669+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.36/docs/secret/scanning/#recommendation for faster secret detection
2023-01-09T20:54:24.575+0100	INFO	Detected OS: debian
2023-01-09T20:54:24.575+0100	INFO	Detecting Debian vulnerabilities...
2023-01-09T20:54:24.586+0100	INFO	Number of language-specific files: 0
```
nor in the latest versions of bitnami/kubectl:
```console
$ trivy image --ignore-unfixed bitnami/kubectl | grep 'CVE-2022-32221'
2023-01-09T20:56:05.666+0100	INFO	Vulnerability scanning is enabled
2023-01-09T20:56:05.666+0100	INFO	Secret scanning is enabled
2023-01-09T20:56:05.666+0100	INFO	If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-09T20:56:05.667+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.36/docs/secret/scanning/#recommendation for faster secret detection
2023-01-09T20:56:09.951+0100	INFO	Detected OS: debian
2023-01-09T20:56:09.951+0100	INFO	Detecting Debian vulnerabilities...
2023-01-09T20:56:09.963+0100	INFO	Number of language-specific files: 1
2023-01-09T20:56:09.963+0100	INFO	Detecting gobinary vulnerabilities...

$ trivy image --ignore-unfixed bitnami/kubectl | grep 'CVE-2021-46848'
2023-01-09T20:56:17.574+0100	INFO	Vulnerability scanning is enabled
2023-01-09T20:56:17.574+0100	INFO	Secret scanning is enabled
2023-01-09T20:56:17.575+0100	INFO	If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-09T20:56:17.575+0100	INFO	Please see also https://aquasecurity.github.io/trivy/v0.36/docs/secret/scanning/#recommendation for faster secret detection
2023-01-09T20:56:19.445+0100	INFO	Detected OS: debian
2023-01-09T20:56:19.445+0100	INFO	Detecting Debian vulnerabilities...
2023-01-09T20:56:19.461+0100	INFO	Number of language-specific files: 1
2023-01-09T20:56:19.461+0100	INFO	Detecting gobinary vulnerabilities...
```

The report you are sharing is from a deprecated bitnami/kubectl version (`1.22-debian-11`) which is expected to contain vulnerabilities since this version is not rebuilt anymore once it was deprecated.

> You mentioned:
> 
> > The Bitnami Application Catalog (OpenSource) is based on Debian 11 but Bitnami, as part of VMware, provides a custom container and Helm Charts catalog based on the desired base image (generic distros such as Debian 10 & 11, CentOS 7, PhotonOS 3 & 4, Ubuntu 18.04, 20.04 & 22.04, RedHat UBI 8 & 9, or custom golden image) through the [VMware Tanzu Application Catalog](https://tanzu.vmware.com/content/blog/use-bitnami-apps-with-vmware-application-catalog).
> 
> Does this mean that the image in the Tanzu catalog hasn't been patched? If so, do you have an ETA on when that might happen?
> 
> Thanks!

No, the image in the [VMware Tanzu Application Catalog](https://tanzu.vmware.com/content/blog/use-bitnami-apps-with-vmware-application-catalog) is not patched but it can be based on different distros that could have included a fix faster than Debian, it depends on the update cadence of each distro. But in the end, we don't apply patches by ourselves, we rely on the update cadence set by each distro.

**Comment by fryz** [comment]
Thanks for the explanations and working through this with me @carrodher - you're the best :) I interact with a handful of open-source image maintainers, and you have provided the best experience I've had thus far. I really appreciate the support you've shown in getting this figured out. 

We've bumped the images that we are using to the `bitnami/kubectl:1.23-debian-11` image, which appears to be supported based on the link you sent earlier. Let me know if that is incorrect. 

Based on my scanning, it does indeed look like CVE-2021-46848 is fixed on this image. Thank you for validating this for me. 

However, it looks like `kubectl:1.23-debian-11` is still affected by CVE-2022-32221: 
```
[10:00:31]$ trivy image bitnami/kubectl:1.23-debian-11 | grep CVE-2022-32221
2023-01-11T10:00:38.186-0500    INFO    Vulnerability scanning is enabled
2023-01-11T10:00:38.186-0500    INFO    Secret scanning is enabled
2023-01-11T10:00:38.186-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-11T10:00:38.186-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-11T10:00:39.292-0500    INFO    Detected OS: debian
2023-01-11T10:00:39.292-0500    INFO    Detecting Debian vulnerabilities...
2023-01-11T10:00:39.310-0500    INFO    Number of language-specific files: 0
│ libcurl3-gnutls   │ CVE-2022-32221   │ CRITICAL │ 7.74.0-1.3+deb11u3    │               │ curl: POST following PUT confusion                           │
```

When I scan the `bitnami/minideb:bullseye` image, it looks like the vulnerability is not present:
```
[10:03:18]$ trivy image bitnami/minideb:bullseye | grep 'CVE-2022-32221'
2023-01-11T10:06:36.139-0500    INFO    Vulnerability scanning is enabled
2023-01-11T10:06:36.139-0500    INFO    Secret scanning is enabled
2023-01-11T10:06:36.139-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-11T10:06:36.139-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-11T10:06:37.510-0500    INFO    Detected OS: debian
2023-01-11T10:06:37.511-0500    INFO    Detecting Debian vulnerabilities...
2023-01-11T10:06:37.531-0500    INFO    Number of language-specific files: 0
```

Finally, what is the significance of the `--ignore-unfixed` flag? From your scan output above, I see that you are using this flag, and when I use this flag I get the same results, but when I exclude this flag I see that the latest kubectl image is vulnerable: 

```
[10:06:37]$ trivy image --ignore-unfixed bitnami/kubectl | grep CVE-2022-32221
2023-01-11T10:10:29.829-0500    INFO    Vulnerability scanning is enabled
2023-01-11T10:10:29.829-0500    INFO    Secret scanning is enabled
2023-01-11T10:10:29.829-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-11T10:10:29.829-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-11T10:10:30.972-0500    INFO    Detected OS: debian
2023-01-11T10:10:30.973-0500    INFO    Detecting Debian vulnerabilities...
2023-01-11T10:10:30.993-0500    INFO    Number of language-specific files: 1
2023-01-11T10:10:30.993-0500    INFO    Detecting gobinary vulnerabilities...

[10:10:31]$ trivy image bitnami/kubectl | grep CVE-2022-32221
2023-01-11T10:10:34.391-0500    INFO    Vulnerability scanning is enabled
2023-01-11T10:10:34.391-0500    INFO    Secret scanning is enabled
2023-01-11T10:10:34.391-0500    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-01-11T10:10:34.391-0500    INFO    Please see also https://aquasecurity.github.io/trivy/v0.32/docs/secret/scanning/#recommendation for faster secret detection
2023-01-11T10:10:35.553-0500    INFO    Detected OS: debian
2023-01-11T10:10:35.553-0500    INFO    Detecting Debian vulnerabilities...
2023-01-11T10:10:35.572-0500    INFO    Number of language-specific files: 1
2023-01-11T10:10:35.572-0500    INFO    Detecting gobinary vulnerabilities...
│ libcurl3-gnutls   │ CVE-2022-32221   │ CRITICAL │ 7.74.0-1.3+deb11u3    │               │ curl: POST following PUT confusion                           │
```

Note that the output from the scans that I included above does NOT use the `--ignore-unfixed` flag. 

Thanks!

**Comment by carrodher** [comment]
Thanks for your kind words 😄 

Let's try to see the different topics:

### The `--ignore-unfixed` flag

When this flag is enabled, the report only displays fixed vulnerabilities, this means there is a new version of the affected package in the distro package manager. If there is any vulnerability reported when using the `--ignore-unfixed` flag means that the container image was not recently rebuilt because there is a new version of the affected package already released in the distro but not included in the image.

In this case, when executing the scanner with this flag, there are no CVEs reported which is the desired situation since it means that, all the packages with a fix available are already included in the container image.

On the other hand, the CVEs reported when this flag is **not** set means there are vulnerable packages but there is not a solution (a new package) provided by the underlying distro so it doesn't matter when the image was rebuilt because it is needed to wait until there is a new release of the package fixing the vulnerability.

### Presence of CVE-2022-32221

This vulnerability reported in the `curl` package is present in the bitnami/kubectl container and there is not any fix provided by the distro (that's why it doesn't appear when the `--ignore-unfixed` is used).
In this case, we can't do anything else apart from waiting until debian and/or the `curl` developers provide a new version to fix the issue.

Taking a look at the [Debian security tracker](https://security-tracker.debian.org/tracker/CVE-2022-32221), we can see how there is already a patch for Debian Bookworm (the testing version), but still no patch for the stable distros. In our case, we are using Bullseye.

#### Why it doesn't appear in `bitnami/minideb`?

This CVE is not reported in `bitnami/minideb` because this base image doesn't include the `curl` package, this package is specifically added to the `bitnami/kubectl` container, see https://github.com/bitnami/containers/blob/b8c7181100bcfed0390e88f83ff454868661c5f0/bitnami/kubectl/1.23/debian-11/Dockerfile#L22

That makes the `curl` package to be installed into the `bitnami/kubectl` container image but not in its base image (`bitnami/minideb`).

**Comment by fryz** [comment]
Right on - once again, thank you for taking the time to walk me through this. 

So it sounds like I need to wait for Debian to fix the vuln and then the kubectl image to be rebuilt and then I should be good to go. 

Should I continue using this issue to keep track of when it will be ready? 

Thanks!

**Comment by carrodher** [comment]
Yep, that should be the process. In the same way, please note that according to the upstream project release cadence, kubectl 1.26 was recently released meaning that 1.23 will be deprecated soon. The supported branches will be 1.24, 1.25 and 1.26.

> Should I continue using this issue to keep track of when it will be ready?

The releases are automated and we don't manually check new versions; when a new version of the affected package is released in the distro, it will be automatically included in the container image, but there is not any notification/track system.
